File: src/ghosts.pandora/app/utils/content_manager.py
import os
import mimetypes
from urllib.parse import unquote
from fastapi import Request
from app_logging import setup_logger
from config.config import STORE_RESULTS
logger = setup_logger(__name__)

class ContentManager:
    def __init__(self, base_dir="_data", default="index", extension="txt"):
        self.base_dir = base_dir
        self.default = default
        self.extension = extension
        self.file_name = None
        self.rel_path = None
        self.full_path = None
        self._store = STORE_RESULTS

    def is_storing(self) -> bool:
        return self._store

    def resolve(self, request: Request):
        self.request = request
        url_path = unquote(request.url.path).lstrip("/")
        
        if url_path.endswith("/"):
            url_path += f"{self.default}.{self.extension}"
        elif not url_path.endswith(f".{self.extension}"):
            url_path += f".{self.extension}"

        self.rel_path = url_path
        self.file_name = os.path.basename(url_path)
        self.full_path = os.path.join(self.base_dir, url_path)
        
        os.makedirs(os.path.dirname(self.full_path), exist_ok=True)

        logger.info(f"{self.extension.upper()} request received. Filename: {self.file_name} Path: {self.rel_path}")

    def guess_media_type(self) -> str:
        media_type, _ = mimetypes.guess_type(self.full_path)
        return media_type or "application/octet-stream"

    def load(self) -> str | None:
        try:
            if not self.full_path or not os.path.exists(self.full_path):
                return None
            with open(self.full_path, "r", encoding="utf-8") as f:
                return f.read()
        except:
            return None

    def save(self, content: str | bytes):
        try:
            os.makedirs(os.path.dirname(self.full_path), exist_ok=True)
            mode = "wb" if isinstance(content, bytes) else "w"
            with open(self.full_path, mode) as f:
                f.write(content)
        except:
            return None

File: src/ghosts.pandora.socializer/src/Infrastructure/Services/ContentManager.cs
using System.Web;

namespace Ghosts.Socializer.Infrastructure.Services;

public class ContentManager
{
    private readonly ApplicationConfiguration _configuration;
    private readonly ILogger<ContentManager> _logger;
    private readonly string _baseDir;
    private readonly string _defaultFileName;
    private readonly string _extension;

    public string FileName { get; private set; }
    public string RelativePath { get; private set; }
    public string FullPath { get; private set; }

    public ContentManager(
        ApplicationConfiguration configuration,
        ILogger<ContentManager> logger,
        string defaultFileName = "index",
        string extension = "txt")
    {
        _configuration = configuration;
        _logger = logger;
        _baseDir = configuration.Pandora.ContentCacheDirectory;
        _defaultFileName = defaultFileName;
        _extension = extension;
    }

    public bool IsStoring()
    {
        return _configuration.Pandora.StoreResults;
    }

    public void Resolve(HttpRequest request)
    {
        var urlPath = HttpUtility.UrlDecode(request.Path.Value?.TrimStart('/') ?? "");

        if (urlPath.EndsWith("/"))
        {
            urlPath += $"{_defaultFileName}.{_extension}";
        }
        else if (!urlPath.EndsWith($".{_extension}"))
        {
            urlPath += $".{_extension}";
        }

        RelativePath = urlPath;
        FileName = Path.GetFileName(urlPath);
        FullPath = Path.Combine(_baseDir, urlPath);

        var directory = Path.GetDirectoryName(FullPath);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }

        _logger.LogInformation("{Extension} request received. Filename: {FileName} Path: {RelativePath}",
            _extension.ToUpper(), FileName, RelativePath);
    }

    public string GuessMediaType()
    {
        var extension = Path.GetExtension(FullPath);
        return extension.ToLower() switch
        {
            ".pdf" => "application/pdf",
            ".doc" => "application/msword",
            ".docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            ".xls" => "application/vnd.ms-excel",
            ".xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            ".ppt" => "application/vnd.ms-powerpoint",
            ".pptx" => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            ".jpg" or ".jpeg" => "image/jpeg",
            ".png" => "image/png",
            ".gif" => "image/gif",
            ".json" => "application/json",
            ".csv" => "text/csv",
            ".txt" => "text/plain",
            ".html" or ".htm" => "text/html",
            ".css" => "text/css",
            ".js" => "application/javascript",
            ".zip" => "application/zip",
            ".tar" => "application/x-tar",
            _ => "application/octet-stream"
        };
    }

    public async Task<byte[]> LoadAsync()
    {
        try
        {
            if (string.IsNullOrEmpty(FullPath) || !File.Exists(FullPath))
            {
                return null;
            }

            return await File.ReadAllBytesAsync(FullPath);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to load cached content from {FullPath}", FullPath);
            return null;
        }
    }

    public async Task SaveAsync(byte[] content)
    {
        try
        {
            var directory = Path.GetDirectoryName(FullPath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            await File.WriteAllBytesAsync(FullPath, content);
            _logger.LogDebug("Saved content to {FullPath}", FullPath);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to save content to {FullPath}", FullPath);
        }
    }

    public async Task SaveAsync(string content)
    {
        try
        {
            var directory = Path.GetDirectoryName(FullPath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            await File.WriteAllTextAsync(FullPath, content);
            _logger.LogDebug("Saved content to {FullPath}", FullPath);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to save content to {FullPath}", FullPath);
        }
    }
}
